File: PythonCodes/Fig09_08.py
#%%
"""
Created on Feb 11 2019
3D surface of normal samples with marginal distributions
@author: Lech A. Grzelak
"""
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st
from mpl_toolkits import mplot3d

def mainCalculation():
    NoOfPaths = 1000
    
    # Parameters

    mu1    = 5.0
    mu2    = 10.0
    sigma1 = 2.0
    sigma2 = 0.5
    rho    = 0.7
    
    # Generation of random samples

    X = np.random.normal(0.0,1.0,[NoOfPaths,1])
    Y = np.random.normal(0.0,1.0,[NoOfPaths,1])
    X = (X - np.mean(X)) / np.std(X)
    Y = (Y - np.mean(Y)) / np.std(Y)
    
    Y = rho *X + np.sqrt(1.0-rho**2) * Y

    # Adjustment for the mean and variance

    X = sigma1* X + mu1
    Y = sigma2* Y + mu2
    
    # 3D plot

    # First we show the scatter plot with the samples

    plt.figure(1)
    ax = plt.axes(projection='3d')
    Z = np.zeros([NoOfPaths,1])
    ax.plot3D(np.squeeze(X), np.squeeze(Y), np.squeeze(Z),'.')
        
    # Now we add the marginal densities

    x1=np.linspace(st.norm.ppf(0.001, loc=mu1, scale=sigma1),st.norm.ppf(1.0-0.001, loc=mu1, scale=sigma1),30)
    y1=np.min(Y) - 1.0 + np.zeros([len(x1),1])
    z1=st.norm.pdf(x1,loc=mu1, scale=sigma1)
    ax.plot3D(np.squeeze(x1), np.squeeze(y1), np.squeeze(z1), 'blue')
    
    x2=np.linspace(st.norm.ppf(0.001, loc=mu2, scale=sigma2),st.norm.ppf(1.0-0.001, loc=mu2, scale=sigma2),30)
    y2=np.min(X) - 1.0 + np.zeros([len(x2),1])
    z2=st.norm.pdf(x2,loc=mu2, scale=sigma2)
    ax.plot3D(np.squeeze(y2), np.squeeze(x2), np.squeeze(z2), 'red')
    
    ax.view_init(50, -120)
mainCalculation()

File: MatlabCodes/Chapter 09/Fig09_08.m
function ScatterPlot3DSurface
clc;close all;

% Monte Carlo model settings

NoOfPaths = 1000; 

mu1    = 5;
mu2    = 10;
sigma1 = 2;
sigma2 = 0.5;
rho    = 0.7;

% Generation of standard normal process

randn('seed',2)
Z1 = random('normal',0,1,[NoOfPaths,1]);
Z2 = random('normal',0,1,[NoOfPaths,1]);
Z1   = (Z1 - mean(Z1)) / std(Z1);
Z2   = (Z2 - mean(Z2)) / std(Z2);

% Correlation 

Z2 = rho*Z1 + sqrt(1-rho^2)*Z2;

% Adjustment for the mean and variance

Z1 = sigma1* Z1 + mu1;
Z2 = sigma2* Z2 + mu2;
        
% Generation of correlated BM, positive correlation

figure1 = figure;

% Create axes

axes1 = axes('Parent',figure1);
figure(1)
hold on;
grid on;
plot(Z1,Z2,'.','linewidth',1.5)
xlabel('X')
ylabel('Y')

% First marginal distribution

x1 =  norminv(0.001,mu1,sigma1)-1:0.1:norminv(1-0.001,mu1,sigma1)+1;
y1 = min(Z2)-1+ zeros([length(x1),1]);
z1 = normpdf(x1,mu1,sigma1);
fill3(x1,y1,z1, 'b', 'FaceAlpha', 1.0)

% Second marginal distribution

x2 =  norminv(0.001,mu2,sigma2)-1:0.1:norminv(1-0.001,mu2,sigma2)+1;
y2 = min(Z1)-1+ zeros([length(x2),1]);
z2 = normpdf(x2,mu2,sigma2);
fill3(y2,x2,z2, 'r', 'FaceAlpha', 1.0)
view(axes1,[119.6 37.2]);
legend('samples','f_X(x)','f_Y(y)')
legend1 = legend(axes1,'show');
set(legend1,...
    'Position',[0.165571430738749 0.758333337590808 0.171428569033742 0.15476190050443]);
